### ISSUE TITLE
[accessibility/frontend] Use client's `prefers-color-scheme` on the web frontend

### ISSUE THREAD

**Issue by tsmethurst** [superseriousbusiness/gotosocial#1214]
_Originally posted by @bremensaki in https://github.com/superseriousbusiness/gotosocial/issues/714#issuecomment-1327991017_

**Comment by benjaminbhollon** [comment]
I'd be happy to look into implementing this; I've done [a fair bit of work with designing accessible CSS](https://readable-css.freedomtowrite.org) and am someone directly affected by the lack of a light mode stylesheet.

If you're looking for an outside contributor on this, let me know; I'll get to work.

**Comment by tsmethurst** [comment]
If you feel like having a crack at it, that would be great! I'm not really sure what the steps involved for this one would be: I myself am clueless about how to have light + dark css in the same stylesheet (?), how to store / modify the user's preference, and how to flip between it in a way that can work without JS. So, input welcome :)

**Comment by benjaminbhollon** [comment]
Okay, got it.

So to kind of outline what I'm imagining:

Without JS, we can use a `@media` query to detect when the browser prefers dark mode:

```
@media (prefers-color-scheme: dark) {
    /* styles go here */
}
```

Then what I can *also* do to add the possibility for users to override that is have the theme switch based on a class or attribute; in my own code I use the `data-theme` attribute set to `light` or `dark` to override the `prefers-color-scheme`.

If that sounds good I can start implementing it.

**Comment by benjaminbhollon** [comment]
Awesome, I'll start looking at the code, then. Haven't worked with sass much (I mostly use plain CSS), so it might take a bit to figure out setting things up.

**Comment by benjaminbhollon** [comment]
Oh, ha, sure! XD

**Comment by tsmethurst** [comment]
> I'm not tobi!

Aren't we all, in a way, tobi? In our hearts? I think so.

Anyway, I think this approach sounds fine. I'd like to hear if @f0x52 has any input too :)

**Comment by VirtualWolf** [comment]
Hey @benjaminbhollon did you end up getting anywhere with this? :)

**Comment by benjaminbhollon** [comment]
I never did, I'm sorry. I think once all the different color schemes came into play (whatever release that was?) I forgot this was something that still needs doing.

**Comment by VirtualWolf** [comment]
All good! I was thinking of taking a look at it and didn't want to step on anyone's toes if it was most of the way to being implemented already.

**Comment by benjaminbhollon** [comment]
My toes are safe, implement away! ;D

**Comment by CDN18** [comment]
@VirtualWolf Thanks for the clarification! I've now submitted #3588, which adds an auto-switch theme version that aligns with the client’s `prefer-color-theme`.

I previously tried using `@import` when merging themes on my own instance, but for some reason, it didn't work. This time it seems fine, so the issue might have been simpler than I thought.

If #3588 meets expectations and gets merged, I think the next step is to create a light version of the base CSS for GoToSocial and add a theme switch button.

**Comment by CDN18** [comment]
With #3352 merged, instance admins can now "enable" `prefers-color-scheme` support on all pages of their instance by loading custom CSS:

I've created an auto-switching version of the color scheme, pasting here in case anyone want to do this:

```css
:root {
	/* Dark Mode Color definitions */
	--white1: #fafaff; /* default text color, contrast >= 5.0 with all $grays */
	--white2: #b3b5c6; /* less important text, can be used with $gray1 (6.8), $gray2 (5.5), $gray3 (4.9), $gray4 (4.5) */

	/* Background shades, contrast >= 5.0 with $white1 (#fafaff) */
	--gray1: #2a2b2f;
	--gray2: #35363b;
	--gray3: #3a3b41;
	--gray4: #45464e;
	--gray5: #4d4e56;
	--gray6: #575861;
	--gray7: #5d5e67;
	--gray8: #696a75;

	--orange1: #fd6a00; /* Used for non-text accent colors, can be used as background: $gray1 for text color (contrast 4.6)*/
	--orange2: #ff853e; /* hover/selected accent to $orange1, can be used with $gray1 (5.7), $gray2 (4.6) */

	--blue1: #3a9fde; /* darker blue for smaller elements (borders), can only be used with $gray1 (4.7) */
	--blue2: #66befe; /* all-round accent color, can be used with $gray1 (6.8), $gray2 (5.5), $gray3 (4.9), $gray4 (4.5) */
	--blue3: #89caff; /* hover/selected accent to $blue2, can be used with $gray1 (7.9), $gray2 (6.3), $gray3 (5.6), $gray4 (5.2), $gray5 (4.7) */

	--error1: #860000; /* Error border/foreground text, can be used with $error2 (5.0), $white1 (10), $white2 (5.1) */
	--error2: #ff9796; /* Error background text, can be used with $error1 (5.0), $gray1 (6.6), $gray2 (5.3), $gray3 (4.8) */
	--error3: #dd2c2c; /* Error button background text, can be used with $white1 (4.51) */
	--error-link: #01318C; /* Error link text, can be used with $error2 (5.56) */

	--green1: #94E749; /* Green for positive/confirmation, similar contrast (luminance) as $blue2 */

	--info-fg: var(--gray1);
	--info-bg: #b3ddff;
	--info-link: var(--error-link);

	--fg: var(--white1);
	--bg: var(--gray1);

	--bg-trans: rgba(77, 78, 86, 0.62);

	--bg-accent: var(--gray5);
	--fg-accent: var(--blue3);
	--fg-reduced: var(--white2);
	--border-accent: var(--orange2);
}

@media (prefers-color-scheme: light) {
	:root {
		/* Light Mode Color definitions */
		--white: #fafaff;  /* we still use white for something */
		--white1: #000000; /* default text color, contrast >= 5.0 with all $grays */
		--white2: #59595F; /* less important text, can be used with $gray1 (8.5) ~ $gray5 (5.0) */

		/* Background shades, contrast >= 5.0 with $white1 (#000000) */
		--gray1: #ffffff;
		--gray2: #f6f6f6;
		--gray3: #ededed;
		--gray4: #e4e4e4;
		--gray5: #dbdbdb;
		--gray6: #d2d2d2;
		--gray7: #c9c9c9;
		--gray8: #c0c0c0;

		--orange1: #c75300; /* Used for non-text accent colors, can be used as background: $gray1 for text color (contrast 4.5)*/
		--orange2: #bd4f00; /* hover/selected accent to $orange1, can be used with $gray1 (5.7), $gray2 (4.5) */

		--blue1: #247cb3; /* lighter blue for smaller elements (borders), can only be used with $gray1 (4.5) */
		--blue2: #166ba0; /* all-round accent color, can be used with $gray1 (5.7), $gray2 (5.3), $gray3 (4.9), $gray4 (4.5) */
		--blue3: #106193; /* hover/selected accent to $blue2, can be used with $gray1 (6.6), $gray2 (6.1), $gray3 (5.6), $gray4 (5.2), $gray5 (4.8) */

		--error1: #ffc1c1; /* Error border/foreground text, can be used with $error2 (5.0), $white1 (13.6), $white2 (4.5) */
		--error2: #aa0000; /* Error background text, can be used with $error1 (5.0), $gray1 (7.7), $gray2 (7.1), $gray3 (6.62) */
		--error3: #d93636; /* Error button background text, can be used with $gray1 ~ $gray4 */
		--error-link: #abd4ff; /* Error link text, can be used with $error2 (5.56) */

		--green1: #367400;

		--info-bg: var(--blue1);
	}
}
```

Since this involves changes to the default behavior, I'm not sure if submitting itself as a PR is appropriate. Maybe I should implement the other feature requests in #2231 and submit them together.

**Comment by tsmethurst** [comment]
A PR with an auto-switching version of the base color scheme would be very welcome indeed if you feel up to making one! :)
